### PR TITLE
Upload attachments v3

### PR DESCRIPTION
**Pull request by vivangkumar**
### What?

Allows attachment uploads using the new v3 api.

### Why?

To finish off multipart messages.

----

### CODE REVIEW

**Comment by vivangkumar** [comment]
[[1] Send v3 message attachment](https://trello.com/c/IL6BLMFt/1645-1-send-v3-message-attachment)

File: Sources/PCMultipartMessage.swift [hunk 1]
```diff
@@ -46,46 +46,134 @@ extension PCMultipartMessage: Hashable {
...
+        self.customData = customData
+        self.size = size
+        self.urlRefresher = urlRefresher
+        self.refreshUrl = refreshUrl
+        self.downloadUrl = downloadUrl
+        self.expiration = expiration
+    }
+
+    func url(completionHandler: @escaping (String?, Error?) -> Void) {
+        if (Date() > self.urlExpiry()) {
```
Comment (callum-oakley) [inline Sources/PCMultipartMessage.swift]: need a 30 minute buffer here

> The url() method on AttachmentPayload must return a URL with at least 30 minutes of validity remaining. If the URL stored in the object has less than 30 minutes of validity remaining, the method must use the refresh URL to update it, and return the new value. The value returned from urlExpiry() must also be updated.

File: Tests/Helpers/Helpers.swift [hunk 2]
```diff
@@ -169,6 +169,28 @@ func deleteInstanceResources(completionHandler: @escaping (TestHelperError?) ->
     }.resume()
 }
 
+func synchronousHTTPRequest(
```
Comment (callum-oakley) [inline Tests/Helpers/Helpers.swift]: If it's synchronous, why not return `(Data?, URLResponse?, Error?)` directly? Having a synchronous method with a completionHandler feels weird.

File: Sources/PCMultipartAttachmentUploader.swift [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        instance: Instance,
+        uploadTasks: [PCMultipartAttachmentUploadTask]
+    ) {
+        self.instance = instance
+        self.uploadTasks = uploadTasks
+    }
+
+    func upload(completionHandler: @escaping ([PCMultipartAttachmentUploadResult]?, [Error]?) -> Void) {
+        for task in self.uploadTasks {
+            self.dispatchGroup.enter()
```
Comment (callum-oakley) [inline Sources/PCMultipartAttachmentUploader.swift]: You could use PCProgressCounter here? Not sure it simplifies things, so up to you.

File: Tests/MessageTests.swift [hunk 4]
```diff
@@ -408,7 +414,510 @@ class MessagesTests: XCTestCase {
...
+                        ex.fulfill()
+                    }
+                }
+            )
+        }
+
+        waitForExpectations(timeout: 15)
+    }
+
+    func testMultipartMessageWithMultipleeAttachmentsRetrievedOnV3() {
```
Comment (callum-oakley) [inline Tests/MessageTests.swift]: `Multiplee`

File: Sources/PCMultipartAttachmentUploader.swift (hunk 3 above)
Comment (vivangkumar) [reply Sources/PCMultipartAttachmentUploader.swift]: Hmmm, I didn't realise this existed :D But I think I'm going to leave it be for now.

File: Tests/Helpers/Helpers.swift (hunk 2 above)
Comment (vivangkumar) [reply Tests/Helpers/Helpers.swift]: d'oh!  🤦‍♂️

File: Sources/PCMultipartMessage.swift [hunk 5]
```diff
@@ -46,46 +46,134 @@ extension PCMultipartMessage: Hashable {
 }
 
 public struct PCPart {
+    let type: String
```
Comment (callum-oakley) [inline Sources/PCMultipartMessage.swift]: Oh, one more thing. This should be part of the payload according to the spec. (where it was before was correct...)

**Comment by vivangkumar** [comment]
@callum-oakley Changes have been made.

File: Sources/PCMultipartMessage.swift [hunk 6]
```diff
@@ -60,32 +64,128 @@ public struct PCMultipartURLPayload {
...
+        self.customData = customData
+        self.size = size
+        self.urlRefresher = urlRefresher
+        self.refreshUrl = refreshUrl
+        self.downloadUrl = downloadUrl
+        self.expiration = expiration
+    }
+
+    func url(completionHandler: @escaping (String?, Error?) -> Void) {
+        if (Date().addingTimeInterval(-(30 * 60)) > self.urlExpiry()) {
```
Comment (callum-oakley) [inline Sources/PCMultipartMessage.swift]:
```suggestion
        if (Date().addingTimeInterval(30 * 60) > self.urlExpiry()) {
```

no?

Comment (vivangkumar) [reply Sources/PCMultipartMessage.swift]: yes 🤦‍♂️ my bad.

**Pull request merged**